contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,368 @@ const (
+domains are available to a StorageClass. To support these administrator
+policies, an AllowedTopology field can also be specified in the
+StorageClass to restrict the topology domains for dynamic provisioning.
+This is not expected to be a common use case, and there are some caveats,
+described below.
+
+```
+type StorageClass struct {
+    ...
+
+    AllowedTopologies *VolumeProvisioningTopologies
+}
+
+type VolumeProvisioningTopologies struct {
+    // Volume must be provisioned from one of the required topologies
+    RequiredTopologies []VolumeTopology
+}
+
+type VolumeTopology struct{
+    // The provisioned volume must support all the specified node labels
+    MatchNodeLabels map[string]string
+}
+```
+
+A nil value means there are no topology restrictions. A scheduler predicate
+will evaluate a non-nil value when considering dynamic provisioning for a node.
+When evaluating the allowed topology against a node, the node needs to have
+labels matching everything specified in a VolumeTopology for one of the
+RequiredTopologies.
+
+A StorageClass update controller can be provided to convert existing topology
+restrictions specified under plugin-specific parameters to the new AllowedTopologies
+field.
+
+The AllowedTopologies will also be provided to provisioners as a new field, detailed in
+the provisioner section.  Provisioners can use the allowed topology information
+in the following scenarios:
+* StorageClass is using the default immediate binding mode. This is the
+  legacy topology-unaware behavior. In this scenario, the volume could be
+  provisioned in a domain that cannot run the Pod since it doesn't take any
+  scheduler input.
+* For volumes that span multiple domains, the AllowedTopologies can restrict those
+  additional domains.  However, special care must be taken to avoid specifying
+  conflicting topology constraints in the Pod. For example, the administrator could
+  restrict a multi-zonal volume to zones 'zone1' and 'zone2', but the Pod could have
+  constraints that restrict it to 'zone1' and 'zone3'.  If 'zone1'
+  fails, the Pod cannot be scheduled to the intended failover zone.
+
+Kubernetes will leave validation and enforcement of the AllowedTopologies content up
+to the provisioner.
+
+TODO: would this be better represented as part of StorageClass quota since that
+is where admins currently specify policies?
+However, ResourceQuota currently only supports quantity, and topology
+restrictions have to be evaluated during scheduling, not admission.
+
+#### Zonal Example
+This example restricts the volumes provisioned to zones us-central1-a and
+us-central1-b.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: zonal-class
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+allowedTopologies:
+  requiredToplogies:
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-a
+  - matchNodeLabels:
```

Comment:
It took me a while to understand why there are two `matchNodeLabels` entries above. It could be useful to explicitly note somewhere that Kubernetes computes OR between matchNodeLabels and AND among individual labels inside.

Reply:
Updated in API comments